### PR TITLE
NFT preparation

### PR DESCRIPTION
**Pull request by arlai-mk**
# Description

Prepares the Hydro and Tribute contracts to the upcoming NFT Marketplace. The primary goal is to make lockups transferable between owners, which requires restructuring storage to avoid having owner addresses in state keys.

#### Lock Storage Redesign
- **Current structure (V1)**: `LOCKS_MAP: key(sender_address, lock_id) -> LockEntry`
  - Owner address is part of the storage key
  - Makes lock transfers difficult as it would require deleting and recreating entries

- **New structure (V2)**: `LOCKS_MAP_V2: key(lock_id) -> LockEntryV2`
  - Lock ID is the only key, making transfers simple
  - Owner address is moved inside the entry value: `LockEntryV2` includes `owner: Addr` field

#### Vote Storage Redesign
- **Current structure**: `VOTE_MAP: key((round_id, tranche_id), sender_addr, lock_id) -> Vote`
  - Owner address is part of the key, binding votes to the original owner

- **New structure (V2)**: `VOTE_MAP_V2: key((round_id, tranche_id), lock_id) -> Vote`
  - Removes address from key, binding votes only to the lock_id
  - Enables accessing the vote without the need to know who voted

#### User Locks Tracking
- Maintains `USER_LOCKS: SnapshotMap<Addr, Vec<u64>>` for efficient tracking of which locks belong to which user
- This map automatically updates when lock ownerships change

#### Tribute Claims Tracking
- **Current structure**: `TRIBUTE_CLAIMS: key(voter_addr, tribute_id) -> amount_claimed`
  - Keeps track whether a user claimed a tribute_id, not sufficient to prevent double-claiming if locks are transferrable
  - Maintained to track historical claim information (`amount_claimed`) of users

- **New additional structure**: `TRIBUTE_CLAIMED_LOCKS: key(tribute_id, lock_id) -> claimed(bool)`
  - Tracks which locks have claimed which tributes at the lock level
  - Enables transferring locks while preventing double-claiming of tributes

### Hydro Contract

- **Lock Management**:
  - Updates functions to use new storage structure
  - Updates `get_past_user_voting_power` to first query `LOCKS_MAP_V2` then `LOCKS_MAP_V1` if locks are not found at requested height
  - Adds ownership validation through a helper function like `get_owned_lock_entry`
  - All functions that operate on locks verify ownership first

- **Vote Management**:
  - Modified to work with `VOTE_MAP_V2` instead of current `VOTE_MAP`
  - Vote casting checks lock ownership
  - Updated vote queries to use lock-based storage

- **Query Functions**:
  - Added a new `QueryMsg::UserVotedLocks` query function that returns detailed lock-level voting data. This query provides a list of all locks a user voted with during a specific round and tranche, including each lock's voting power. While the existing `QueryMsg::UserVotes` aggregates total voting power by proposal, this new query explicitly identifies which individual locks were used for voting. The tribute contract now uses this information during the tribute claims, as well as the V2 migration process (to properly track which locks have claimed which tributes).
  - Updated existing queries to work with new storage structure

### Tribute Contract

- **Tribute Claiming**:
  - Modified to work with lock-based approach
  - Check if a lock has already claimed a tribute
  - Records claims in both old (for historical data: amount claimed by user) and new (for duplicate claiming prevention) storage

- **Query Functions**:
  - Updated `OutstandingTributeClaims` to check lock-specific claim status
  - Added lock information to tribute claim responses

## 3. Migration Strategy
The migration to V2 storage requires a migration of Hydro contract first, then Tribute contract.

### Hydro Contract
The migration to V2 storage occurs in a new migration module:

1. **Lock Entries Migration**:
   - Read all lock entries from current `LOCKS_MAP`
   - For each entry, create a new `LockEntryV2` with owner information
   - Save new entries to `LOCKS_MAP_V2` (starting at height of the migration)
   - Preserve original entries in `LOCKS_MAP_V1` for historical queries, pre-migration (as in `get_past_user_voting_power`)

2. **Vote Migration**:
   - Read all votes from current `VOTE_MAP`
   - Save to `VOTE_MAP_V2` with lock_id as key instead of (address, lock_id)
   - `VOTE_MAP_V1` would only be preserved temporarily during migration
   - After migration, only `VOTE_MAP_V2` would be used

3. **Migration process**
To handle potential gas limit constraints, the migration process has been split into manageable batches using these message types:
- `MigrateMsg::MigrateLocksV1ToV2 { start, limit }`: Migrates a batch of locks
- `MigrateMsg::MigrateVotesV1ToV2 { start, limit }`: Migrates a batch of votes

When the contract processes either migration message type, it:
1. Puts the hydro contract in pause mode, which prevents any other contract operations
2. Processes the requested batch of data
3. Only unpause the contract when both lock and vote migrations are complete

### Tribute Contract
The migration handles retrofitting existing tribute claims:

1. **Claimed Tribute Migration**:
   - Group existing tribute claims by user/round/tranche for optimization
   - For each group (called "context"), queries Hydro for locks that voted for proposals using `UserVotedLocks`
   - Creates new entries in `TRIBUTE_CLAIMED_LOCKS` for each lock_id that voted for a proposal with a claimed tribute

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [x] Targeted the correct branch
* [x] Included the necessary unit tests
* [ ] Added/adjusted the necessary [interchain tests](./test/interchain/)
* [x] Added a changelog entry in `.changelog`
* [x] Compiled the contracts by using `make compile` and included content of the *artifacts* directory into the PR
* [x] Regenerated front-end schema by using `make schema` and included generated files into the PR
* [ ] Updated the relevant documentation or specification
* [x] Reviewed "Files changed" and left comments if necessary
* [ ] Confirmed all CI checks have passed

### CODE REVIEW

**Review by p-offtermatt — COMMENTED** [review]
I think all the deprecated structured should probably more clearly be deprecated. the v1/v2 in the main Hydro contract works, but in the tribute contract it seems a bit subtle

File: contracts/hydro/src/state.rs [hunk 1]
```diff
@@ -100,22 +100,53 @@ pub const LOCK_ID: Item<u64> = Item::new("lock_id");
 // this is incremented every time a new proposal is created
 pub const PROP_ID: Item<u64> = Item::new("prop_id");
 
-// LOCKS_MAP: key(sender_address, lock_id) -> LockEntry
-pub const LOCKS_MAP: SnapshotMap<(Addr, u64), LockEntry> = SnapshotMap::new(
+// LOCKS_MAP_V1: Previous structure, now preserved for migration, and for historical queries
```
[the remark this answers is not in the document]
Comment (arlai-mk) [reply contracts/hydro/src/state.rs:103]: Basically, once the contract is migrated, the old struct is used in 2 places:
- `can_user_lock_in_known_users_cap`: the `round_to_check` can only be `current_round - 1`, so in this case we don't need to query the historical VP for before previous round.
- `query_voting_power_at_height`: unless some frontend requires access to old voting power, it is currently only used by DAO governance voting adapter to know how much voting power a user had at the start of a DAODAO proposal. Once a DAODAO proposal is executed, it should not need to access past users voting power anymore.

As voting power decays, and currently fully after 3 rounds at most, and we may need to query voting power in previous round, then I suppose it would be possible to remove LOCKS_MAP_V1 after 4 full rounds after migration.

Let's say migration happens during round 6, then from round 11 onwards, it should be safe to delete the old struct from the store (as long as we don't have other needs for historical VP info).

File: contracts/hydro/src/contract.rs [hunk 2]
```diff
@@ -604,38 +595,33 @@ fn unlock_tokens(
...
+        .collect::<Result<_, _>>()?;
 
-    let mut to_delete = vec![];
     let mut total_unlocked_amount = Uint128::zero();
 
     let mut response = Response::new()
         .add_attribute("action", "unlock_tokens")
         .add_attribute("sender", info.sender.to_string());
 
+    let mut removed_lock_ids = HashSet::new();
```
[the remark this answers is not in the document]
Comment (arlai-mk) [reply contracts/hydro/src/contract.rs:620]: Yes, basically we took the previously named `to_delete` which was also basically the same as `unlocked_lock_ids`, although one is a `vec` of `u64` and one a `vec` of `string`.

`to_delete` was then converted to a `HashSet` of `u64`, so we just skipped the conversion by creating `removed_lock_ids` as a `HashSet` from the start.
I suppose we could just optimize it a bit and keep only one of both `removed_lock_ids` and `unlocked_lock_ids`. Let me know if you want and I'll update it.

[the remark this answers is not in the document]
Comment (p-offtermatt) [reply contracts/hydro/src/contract.rs:695]: Seems cleaner to me to only have one, yep

File: contracts/hydro/src/utils.rs [hunk 3]
```diff
@@ -563,13 +574,46 @@ pub struct LockingInfo {
...
+    // Verify sender is the owner
+    if lock_entry.owner != *sender {
+        return Err(ContractError::Std(StdError::generic_err(format!(
+            "Lock with ID {} is not owned by {}",
+            lock_id, sender
+        ))));
+    }
+
+    Ok(lock_entry)
+}
```
[the remark this answers is not in the document]
Comment (arlai-mk) [reply contracts/hydro/src/utils.rs:601]: hmm, you're right, it seems a duplicate of `get_owned_lock_entry`, which is used.
Apologies for this, I suppose Victor any myself wrote 2 versions of the same function, and I didn't notice this one. Will remove.

File: contracts/hydro/src/state.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (arlai-mk) [reply contracts/hydro/src/state.rs:107]: I added a comment on top of the LOCKS_MAP_V1 structure:
> // Note: as discussed in https://github.com/informalsystems/hydro/pull/244#discussion_r2065764776,
> //  if the upgrade goes live during round 7, it could be safely removed in round 12

File: contracts/hydro/src/contract.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (arlai-mk) [reply contracts/hydro/src/contract.rs:695]: I removed `unlocked_lock_ids` and kept the `removed_lock_ids` as `HashSet<u64>` which is used for efficient lookups in the `retain` operation when updating `USER_LOCKS`.

Then, I convert it to `unlocked_lock_ids` at the end to be used as attributes.
https://github.com/informalsystems/hydro/blob/51f2cc0a7fe24e349001d38ad8a9e5a597f26046/contracts/hydro/src/contract.rs#L744-L748

File: contracts/tribute/src/contract.rs [hunk 4]
```diff
@@ -578,6 +621,7 @@ pub fn query_round_tributes(
 // then checks whether the given user address can claim them.
 // If the user has not claimed the tribute yet, the amount that the user would receive when claiming is
 // computed, and the tribute is added to the list of tributes that the user can claim.
+// TODO: Do we still want start_from and limit? Seems not convenient in the new situation?
```
[the remark this answers is not in the document]
Comment (arlai-mk) [reply contracts/tribute/src/contract.rs]: I removed `start_from` and `limit` from `query_round_tributes`. 
Please note that it affects the schema, and so clients (frontends) using `QueryMsg::OutstandingTributeClaims` will need to update their types.
Also, for information, I had to remove one test that was setup with out of range `start_from`

File: contracts/hydro/src/utils.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (arlai-mk) [reply contracts/hydro/src/utils.rs]: Removed

**Review by p-offtermatt — APPROVED** [review]
